adodb-lib.inc.php

```diff
@@ -1208,64 +1249,156 @@ function _adodb_debug_execute($zthis, $sql, $inputarr)
 		if($emsg = $zthis->ErrorMsg()) {
 			if ($err = $zthis->ErrorNo()) {
 				if ($zthis->debug === -99)
-					ADOConnection::outp( "<hr>\n($dbt): ".htmlspecialchars($sqlTxt)." &nbsp; $ss\n<hr>\n",false);
+					
+					ADOConnection::outp( "<hr>\n($myDatabaseType): ".htmlspecialchars($sqlTxt)." &nbsp; $ss\n<hr>\n",false);
 
 				ADOConnection::outp($err.': '.$emsg);
 			}
 		}
-	} else if (!$qID) {
-
+	} 
+	else if (!$qID) {
+		/*
+		* Statement execution has failed
+		*/
 		if ($zthis->debug === -99)
-				if ($inBrowser) ADOConnection::outp( "<hr>\n($dbt): ".htmlspecialchars($sqlTxt)." &nbsp; $ss\n<hr>\n",false);
-				else ADOConnection::outp("-----<hr>\n($dbt): ".$sqlTxt."$ss\n-----<hr>\n",false);
-
+		{
+			if ($inBrowser)
+			{
+				$outString = "<hr class='adodb-debug'>(%s):  %s &nbsp; %s<hr class='adodb-debug'>";
+				ADOConnection::outp(sprintf($outString, $myDatabaseType,htmlspecialchars($sqlTxt),$ss),false);
+			}
+			else
+			{				
+				$outString = sprintf("%s\n%s\n    %s\n%s\n",str_repeat('-',78),$myDatabaseType,$sqlTxt,$ss,str_repeat('-',78));
+				ADOConnection::outp($outString,false);
+			}
+		}
+		
+		/*
+		* Send last error to output
+		*/
 		ADOConnection::outp($zthis->ErrorNo() .': '. $zthis->ErrorMsg());
 	}
 
-	if ($zthis->debug === 99) _adodb_backtrace(true,9999,2);
+	if ($zthis->debug === 99) 
+	{
+		_adodb_backtrace(true,9999,2);
+	}
```

Comment:
The backtrace is printed twice in 99 mode, because adodb_backtrace() is already being called in ADOConnection::execute().

```
------------------------------------------------------------------------------
mysqli-
    select * from xxx where id = ? 
[ (0=>'i') (1=>'1')  ]
SQL Statement failed on preparation: select * from xxx where id = ?'
1146: Table 'bugtracker.xxx' doesn't exist

### This is the trace from adodb-lib.inc.php line 1285
			ADOConnection.Query(select * from xxx where id = ?, Array[1])% line   16, file: /home/dregad/dev/adodb/t3.php
		test(select * from xxx where id = ?, Array[1])% line    7, file: /home/dregad/dev/adodb/t3.php

### This is the trace from adodb.inc.php line 1586
			ADOConnection._Execute(select * from xxx where id = ?, Array[2])% line 1144, file: /home/dregad/dev/adodb/drivers/adodb-mysqli.inc.php
		ADODB_mysqli.execute(select * from xxx where id = ?, Array[1])% line 1219, file: /home/dregad/dev/adodb/adodb.inc.php
	ADOConnection.Query(select * from xxx where id = ?, Array[1])% line   16, file: /home/dregad/dev/adodb/t3.php
test(select * from xxx where id = ?, Array[1])% line    7, file: /home/dregad/dev/adodb/t3.php
```

I would suggest removing the call in ADOConnection::execute() (see https://github.com/ADOdb/ADOdb/blob/master/adodb.inc.php#L1585-L1587), and handle the backtrace output in _adodb_debug_execute(). Also, I don't understand why we're ignoring 2 elements in the backtrace. Feels like trying to do what is already taken care of by the array_shift() calls in _adodb_backtrace(). So:
```php
if ($qID === false || $zthis->debug === 99) {
    _adodb_backtrace();
}
```